File: 2025-10-October-GFG-POTD/October 24 - Split Array Subsequences/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool isPossible(vector<int>& arr, int k) {
        unordered_map<int, int> count, endAt;

        // Step 1: Count frequency of each number
        for (int num : arr) {
            count[num]++;
        }

        // Step 2: Traverse and form subsequences
        for (int num : arr) {
            if (count[num] == 0) continue;

            // Case 1: Extend an existing subsequence ending at num - 1
            if (endAt[num - 1] > 0) {
                endAt[num - 1]--;
                endAt[num]++;
                count[num]--;
            } 
            // Case 2: Start a new subsequence of length k
            else {
                bool possible = true;
                for (int i = 0; i < k; i++) {
                    if (count[num + i] <= 0) {
                        possible = false;
                        break;
                    }
                }
                if (!possible) return false;

                for (int i = 0; i < k; i++) {
                    count[num + i]--;
                }
                endAt[num + k - 1]++;
            }
        }

        return true;
    }
};

// Example usage
int main() {
    Solution s;
    vector<int> arr1 = {2,2,3,3,4,5};
    vector<int> arr2 = {1,1,1,1,1};
    cout << boolalpha;
    cout << s.isPossible(arr1, 2) << endl; // true
    cout << s.isPossible(arr2, 4) << endl; // false
}

File: 2025-10-October-GFG-POTD/October 24 - Split Array Subsequences/Solution.java
class Solution {

    public boolean isPossible(int[] arr, int k) {
        Map<Integer, Integer> count = new HashMap<>();
        Map<Integer, Integer> endAt = new HashMap<>();

        // Step 1: Count frequency of each number
        for (int num : arr) {
            count.put(num, count.getOrDefault(num, 0) + 1);
        }

        // Step 2: Traverse and form subsequences
        for (int num : arr) {
            if (count.get(num) == 0) continue;

            // Case 1: Extend an existing subsequence ending at num - 1
            if (endAt.getOrDefault(num - 1, 0) > 0) {
                endAt.put(num - 1, endAt.get(num - 1) - 1);
                endAt.put(num, endAt.getOrDefault(num, 0) + 1);
                count.put(num, count.get(num) - 1);
            }
            // Case 2: Start a new subsequence of length k
            else {
                boolean possible = true;
                for (int i = 0; i < k; i++) {
                    if (count.getOrDefault(num + i, 0) <= 0) {
                        possible = false;
                        break;
                    }
                }
                if (!possible) return false;

                for (int i = 0; i < k; i++) {
                    count.put(num + i, count.get(num + i) - 1);
                }
                endAt.put(num + k - 1, endAt.getOrDefault(num + k - 1, 0) + 1);
            }
        }

        return true;
    }
}
